scripts/zkai_native_attention_mlp_source_backed_adapter_selector_gate.py

```
@@ -0,0 +1,566 @@
+    refresh_payload_commitment(payload)
+    validate_payload(payload, context=context)
+    payload["mutation_result"] = mutation_result(payload, context)
+    refresh_payload_commitment(payload)
+    validate_payload(payload, context=context)
+    return payload
+
+
+def validate_payload(payload: dict[str, Any], *, context: dict[str, Any] | None = None) -> None:
+    context = build_context() if context is None else context
+    if set(payload) != FINAL_KEYS:
+        raise SourceBackedAdapterSelectorError("top-level key drift")
```

Ruff (0.15.12)

[warning] Avoid specifying long messages outside the exception class

(TRY003)